ports/stm32h7/board_flash.c

```diff
@@ -0,0 +1,228 @@
+
+#if BOARD_SPI_FLASH_EN
+SPI_HandleTypeDef _spi_flash;
+#endif // BOARD_SPI_FLASH_EN
+
+//--------------------------------------------------------------------+
+// Board Memory Callouts
+//--------------------------------------------------------------------+
+#ifdef W25Qx_SPI
+uint32_t W25Qx_SPI_Transmit(uint8_t * buffer, uint16_t len, uint32_t timeout)
+{
+  return (uint32_t) HAL_SPI_Transmit(&_spi_flash, buffer, len, timeout);
+}
+
+uint32_t W25Qx_SPI_Receive(uint8_t * buffer, uint16_t len, uint32_t timeout)
+{
+  return (uint32_t) HAL_SPI_Receive(&_spi_flash, buffer, len, timeout);
+}
+
+void W25Qx_Delay(uint32_t ms)
+{
+  HAL_Delay(ms);
+}
+
+uint32_t W25Qx_GetTick(void)
+{
+  return HAL_GetTick();
+}
+#endif // W25Qx_SPI
+
+//--------------------------------------------------------------------+
+// Flash LL for tinyuf2
+//--------------------------------------------------------------------+
+
+
+extern volatile uint32_t _board_tmp_boot_addr[];
+extern volatile uint32_t _board_tmp_boot_magic[];
+
+#define TMP_BOOT_ADDR   _board_tmp_boot_addr[0]
+#define TMP_BOOT_MAGIC  _board_tmp_boot_magic[0]
+
+uint32_t board_get_app_start_address(void)
+{
+  if (TMP_BOOT_MAGIC == 0xDEADBEEFU) 
+  {
+    return TMP_BOOT_ADDR;
+  }
+  else 
+  {
+    return BOARD_QSPI_APP_ADDR;
+  }
+}
+
+void board_save_app_start_address(uint32_t addr)
+{
+  TMP_BOOT_MAGIC = 0xDEADBEEFU;
+  TMP_BOOT_ADDR = addr;
+}
+
+void board_clear_temp_boot_addr(void)
+{
+  TMP_BOOT_MAGIC = 0x00U;
+  TMP_BOOT_ADDR = 0x00U;
+}
+
+void board_flash_early_init(void)
+{
+#if BOARD_QSPI_FLASH_EN
+  // QSPI is initialized early to check for executable code
+  qspi_flash_init(&_qspi_flash);
+  // Initialize QSPI driver
+  w25qxx_Init();
+  // SPI -> QPI
+  w25qxx_EnterQPI();
+#endif // BOARD_QSPI_FLASH_EN
+}
+
+void board_flash_init(void)
+{
+#if BOARD_SPI_FLASH_EN
+  // Initialize SPI peripheral
+  spi_flash_init(&_spi_flash);
+  // Initialize SPI drivers
+  W25Qx_Init();
+#endif // BOARD_SPI_FLASH_EN
+}
+
+void board_flash_deinit(void)
+{
+#if BOARD_QSPI_FLASH_EN
+  // Enable Memory Mapped Mode
+  // QSPI flash will be available at 0x90000000U (readonly)
+  w25qxx_Startup(w25qxx_DTRMode);
+#endif // BOARD_QSPI_FLASH_EN
+}
+
+uint32_t board_flash_size(void)
+{
+  // TODO: how do we handle more than 1 target here?
+  return 8*1024*1024;
+}
+
+void board_flash_flush(void)
+{
+  // TODO: do we need to implement this if there no caching involved?
+  // maybe flush cached RAM here?
+}
+
+void board_flash_read(uint32_t addr, void * data, uint32_t len)
+{
+  TUF2_LOG1("Reading %lu byte(s) from 0x%08lx\r\n", len, addr);
+#if BOARD_QSPI_FLASH_EN
+  // addr += QSPI_BASE_ADDR;
+  if (IS_QSPI_ADDR(addr))
+  {
+    (void) W25qxx_Read(data, addr - QSPI_BASE_ADDR, len);
+    return;
+  }
+#endif
+
+#if BOARD_AXISRAM_EN
+  if (IS_AXISRAM_ADDR(addr) && IS_AXISRAM_ADDR(addr + len - 1))
+  {
+    memcpy(data, (void *) addr, len);
+    return;
+  }
+#endif // BOARD_AXISRAM_EN
+
+  if (IS_PFLASH_ADDR(addr))
```

Comment:
Missing SPI check here, only doing QSPI/RAM/InternalFlash
I found this while playing around with the code to try and get my .uf2 to run...
Will try and double-check for similar missing pieces as soon as i get this working...

Reply:
`board_flash_read` is used to provide data for the `CURRENT.uf2`, but it's not really possible to target multiple memories for a readout. (maybe if we readout internal/QSPI/SPI flash and then concatenate all data into CURRENT.uf2)
Internally within the port it's used to read the vector table, and the SPI flash cannot contain executables, so I left it out of `board_flash_read`.